### ISSUE TITLE
ICML paragraph styles ending with "> Paragraph"

### ISSUE THREAD

**Issue by ptram** [jgm/pandoc#8333]
Hi,

I see that paragraph styles are converted with a "> Paragraph" extension added to the original paragraph style name.

Is there a reason for this redundant information? I find it a bit of an annoyance, since it will break the stylesheet, and make reading the name in the style palette more difficult.

Is there a way to avoid this extension to be added, maybe with a command passed to Pandoc together with the other attributes?

Paolo

**Comment by iandol** [comment]
Here is an example:

```
▶︎ pandoc -t icml
::: {custom-style="Poetry"}
| A Bird came down the Walk---
| He did not know I saw---
:::
            
<ParagraphStyleRange AppliedParagraphStyle="ParagraphStyle/Poetry &gt; Paragraph">
  <CharacterStyleRange AppliedCharacterStyle="$ID/NormalCharacterStyle">
    <Content>A Bird came down the Walk—</Content>
  </CharacterStyleRange>
  <CharacterStyleRange AppliedCharacterStyle="$ID/NormalCharacterStyle">
    <Content>&#x2028;</Content>
  </CharacterStyleRange>
  <CharacterStyleRange AppliedCharacterStyle="$ID/NormalCharacterStyle">
    <Content>He did not know I saw—</Content>
  </CharacterStyleRange>
</ParagraphStyleRange>
```

**Comment by ptram** [comment]
> Here is an example:

Exactly. Is there a way to avoid the `&gt; Paragraph` part to be automatically added to the name of the paragraph style?

Paolo

**Comment by iandol** [comment]
Not that I know of, I think the line that does this is this one?

https://github.com/jgm/pandoc/blob/master/src/Text/Pandoc/Writers/ICML.hs#L527

I don't know any haskell and can't grok why it appends ` > Paragraph`. As a quick workaround you can use a text editor or script and remove this fragment.

**Comment by jgm** [comment]
I don't know a thing about ICML. @mb21 wrote the ICML writer, if I recall, and might be able to explain why this is here.

**Comment by ptram** [comment]
> If you want to differentiate between normal paragraphs, and paragrphas inside your Poetry block, you need two different style names.

I understand. My objection is that this is not always needed (to be true: it is very rarely needed), and can be done at the origin. Automatically changing the original style names scrambles the stylesheet, and makes the use of Pandoc as an interchange format more complicate.

Take for example this case: You are conforming to a company or publishing house style. You write/edit with those styles. When converting to ICML, you get the paragraph style names changed.

At this point, you have to manually edit all the converted ICML files and all the styles, to avoid breaking compatibility with the house style.

Paolo

**Comment by ptram** [comment]
By the way – sorry if this should be clear to most, but I'm new to the conversion filters and I'm not a coder. Is there a way, on my side, to use a customized version of the "Text.Pandoc.Writers.ICML" module, to try some changes? Are there instructions on who to install it in my Mac (Mojave)?

Paolo

**Comment by jgm** [comment]
https://pandoc.org/installing.html#compiling-from-source

**Comment by ptram** [comment]
> https://pandoc.org/installing.html#compiling-from-source

Thank you, John. Let's see if I can dirty my fingers with the compilers, without cutting them!

Paolo

**Comment by mb21** [comment]
> Automatically changing the original style names scrambles the stylesheet

Not sure I understand your workflow. Your input format is markdown? Can you post an example?

Your problem is specific to the `custom-style` attribute?

Try the following example, and you'll see why we need to generate style names and you'll have to adjust your company's template if you want your input to be markdown anyway:

```markdown
## title

my para

> my blockquote

::: {custom-style="Aside"}
## Title in the aside

my para in the aside

> my blockquote in the aside
:::
```

**Comment by ptram** [comment]
> Try the following example

I get this result:

<img width="567" alt="image" src="https://user-images.githubusercontent.com/19186065/193247066-5c84ee9e-d627-40a5-b387-590eb818b585.png">

What is not working, here, for me, is that the systems decides the style names taking the place of the author. A stylesheet is always pre-existing, inherited either from the publishing house or previous projects from which you are reusing some parts.

A stylesheet always includes style names. If the converter pretends to create different ones, you end up with something that is in practice impossible to edit, due to the high number of styles to be manually renamed in the various files of a project.

This is a snippet from document made in Scrivener, from an original in InDesign, and converted to Pandoc markdown:

```
::: {custom-style="Heading 3"}
Heading 3 paragraph
:::

::: {custom-style="Image"}
{image.tiff}
:::
```

In the original document, I have the "Heading 3" and "Image" paragraph styles applied to these paragraphs.

The Pandoc-->icml converter changes the names into "Heading 3 > Paragraph" and "Image > Paragraph". They no longer match the stylesheet. Everything has to be applied manually line by line, without being able to let InDesign match the incoming text with the template's styles.

Paolo

**Comment by mb21** [comment]
> What is not working, here, for me, is that the system decides the style names

I see. Then you need to either post-process the output (with something like `sed`), or use a different tool. 

Wrapping some thing in divs like you did in your example (the `:::` syntax) adds an additional element to the pandoc document AST (you can see this better with `-t html` or `-t native`), it's not renaming the style of the element its wrapping. That's because it's intended for things like `aside` boxes or similar, that can contain arbitrary content.

**Comment by ptram** [comment]
> Wrapping some thing in divs like you did in your example (the `:::` syntax) adds an additional element to the pandoc document AST (you can see this better with `-t html` or `-t native`), it's not renaming the style of the element its wrapping. That's because it's intended for things like `aside` boxes or similar, that can contain arbitrary content.

For what I can read in the other thread you linked above (containing a detailed explanation of how creating custom styles work), it seems to me that fencing is mandatory for creating new paragraph styles.

Is there an alternative way to create them, without considering them included into containers like boxes or asides?

Post-processing the generated ICML file is a possibility, but also one passage more, that would slow the whole process and open a possibility for errors. It would be great if it wouldn't be needed, and the converter could also simply generated a code with the same styles named as in the Pandoc file.

Paolo

**Comment by mb21** [comment]
> fencing is mandatory for creating new paragraph styles.

that's correct. because currently not all pandoc AST elements support custom attributes.

**Comment by dmoles** [comment]
FWIW, I think these styles have two purposes:

1. to differentiate between:
   1.  simple lists where each item is just a string of text, and
   2. more complex lists where each item is one or more paragraphs
2. to work around the fact that ICML styles aren't sophisticated enough for some of the things people often want to do in HTML / CSS

If you take a fairly simple markdown doc like this and run it through Pandoc:

```markdown
list 1:

- a
- simple
- list

list 2:

- a

- paragraph

  subpara

- list
``` 

You get this HTML (comments added):

```html
<!-- list 1:     --> <p>list 1:</p>
<ul>
<!-- - a         --> <li>a</li>               <!--  note each item contains only text -->
<!-- - simple    --> <li>simple</li>
<!-- - list      --> <li>list</li>
</ul>
<!-- list 2:     --> <p>list 2:</p>
<ul>
<!-- - a         --> <li><p>a</p></li>       <!--  note each item contains one or more paragraphs -->
<!-- - paragraph --> <li><p>paragraph</p>
<!--   subpara  --> <p>subpara</p></li>
<!-- - list      --> <li><p>list</p></li>
</ul>
```

ICML doesn't have a concept of lists as first-class objects in the same way as HTML, so what you get instead (omitting the paragraph content and the `<Br />` elements) is:

```xml
<!-- list 1:     --> <ParagraphStyleRange AppliedParagraphStyle="ParagraphStyle/Paragraph">
<!-- - a         --> <ParagraphStyleRange AppliedParagraphStyle="ParagraphStyle/BulList &gt; first" NumberingContinue="false">
<!-- - simple    --> <ParagraphStyleRange AppliedParagraphStyle="ParagraphStyle/BulList">
<!-- - list      --> <ParagraphStyleRange AppliedParagraphStyle="ParagraphStyle/BulList">
<!-- list 2:     --> <ParagraphStyleRange AppliedParagraphStyle="ParagraphStyle/Paragraph">
<!-- - a         --> <ParagraphStyleRange AppliedParagraphStyle="ParagraphStyle/BulList &gt; first &gt; Paragraph" NumberingContinue="false">
<!-- - paragraph --> <ParagraphStyleRange AppliedParagraphStyle="ParagraphStyle/BulList &gt; Paragraph">
<!--   subpara  --> <ParagraphStyleRange AppliedParagraphStyle="ParagraphStyle/BulList &gt; subParagraph &gt; Paragraph">
<!-- - list      --> <ParagraphStyleRange AppliedParagraphStyle="ParagraphStyle/BulList &gt; Paragraph">
```

In HTML, `<li>` is a block element, so you can put multiple `<p>` elements inside it, and you'll only get one
actual bullet character. In ICML, there's no list block element, so if you want to have multiple paragraphs under a single bullet, you need different styles, so one (`BulList > Paragraph`) gets the bullet, and the other (`BulList > subParagraph > Paragraph`) doesn't.

Or say you want to have extra space before each list. In HTML/CSS, you can set the top margin on the `<ul>` itself, but in ICML, again, there is no list block element. So you have to make the first list element a different paragraph style (`BulList > first`) and put the extra space before on that. (Of course you might also want space *after*, and Pandoc doesn't give you a `> last` out of the box, but oh well.)

The same goes for more complex situations like:

```markdown
- bulleted lists with
  1. numbered lists
  2. inside
     1. and multileveled
     2. lists
        - of all kinds

> 1. or lists
> 2. inside
>    - blockquotes
```

There's no very good way to represent these in ICML (that I've found, anyway, but I haven't really dug through [the specification](https://web.archive.org/web/20220218040046/https://www.adobe.com/content/dam/acom/en/devnet/indesign/sdk/cs6/idml/idml-specification.pdf) yet), so Pandoc's workaround is an explosion of paragraph styles.
